core/app/models/spree/payment_create.rb

```diff
@@ -0,0 +1,54 @@
+module Spree
+  class PaymentCreate
+    def initialize(order, attributes, payment: nil, request_env: {})
+      @order = order
+      @payment = payment
+      @attributes = attributes.dup
+      @source_attributes = attributes.delete(:source_attributes) || {}
+      @request_env = request_env
+    end
+
+    def build
+      @payment ||= order.payments.new
+      @payment.request_env = @request_env if @request_env
+      @payment.attributes = @attributes
+
+      if source_attributes[:existing_card_id].present?
+        build_existing_card
+      else
+        build_source
+      end
+
+      @payment
+    end
+
+    private
+
+    attr_reader :order, :payment, :attributes, :source_attributes
+
+    def build_source
+      payment_method = payment.payment_method
+      if source_attributes.present? && payment_method.try(:payment_source_class)
+        payment.source = payment_method.payment_source_class.new(source_attributes)
+        payment.source.payment_method_id = payment_method.id
+        payment.source.user_id = order.user_id if order
+      end
+    end
+
+    def build_existing_card
+      credit_card = CreditCard.
+        where(user_id: order.user_id).
+        where.not(user_id: nil).
```

Comment:
You don't need to check if `user_id is not null`. If user is null for order, then CreditCard will be not found, so `find` will throw exception.

In case when user_id is null, you can just stop executing code and throw exception. I don't know if there could be order without user at all.